src/read_write/node_iterator.rs

```diff
@@ -2,178 +2,192 @@ use crate::codec::{decode, fixpoint_decode};
-                point.position.z =
-                    fixpoint_decode(iter.xyz_reader.read_u8().unwrap(), min.z, edge_length);
-            }
-            PositionEncoding::Uint16 => {
-                point.position.x = fixpoint_decode(
-                    iter.xyz_reader.read_u16::<LittleEndian>().unwrap(),
+            PositionEncoding::Float64 => {
+                point.position.x = decode(
+                    self.readers[0].read_f64::<LittleEndian>().unwrap(),
                     min.x,
                     edge_length,
                 );
-                point.position.y = fixpoint_decode(
-                    iter.xyz_reader.read_u16::<LittleEndian>().unwrap(),
+                point.position.y = decode(
+                    self.readers[0].read_f64::<LittleEndian>().unwrap(),
                     min.y,
                     edge_length,
                 );
-                point.position.z = fixpoint_decode(
-                    iter.xyz_reader.read_u16::<LittleEndian>().unwrap(),
+                point.position.z = decode(
+                    self.readers[0].read_f64::<LittleEndian>().unwrap(),
                     min.z,
                     edge_length,
                 );
             }
         }
 
-        point.color.red = iter.rgb_reader.read_u8().unwrap();
-        point.color.green = iter.rgb_reader.read_u8().unwrap();
-        point.color.blue = iter.rgb_reader.read_u8().unwrap();
+        point.color.red = self.readers[1].read_u8().unwrap();
+        point.color.green = self.readers[1].read_u8().unwrap();
+        point.color.blue = self.readers[1].read_u8().unwrap();
 
-        point.intensity = iter
-            .intensity_reader
-            .as_mut()
+        point.intensity = self
+            .readers
+            .get_mut(2)
             .map(|ir| ir.read_f32::<LittleEndian>().unwrap());
+
+        point
+    }
+}
+
+impl CubeNodeReader {
+    pub fn new(
+        mut layers: HashMap<NodeLayer, Box<dyn Read>>,
+        position_encoding: PositionEncoding,
+        bounding_cube: Cube,
+    ) -> Result<Self> {
+        let xyz_reader = BufReader::new(
+            layers
+                .remove(&NodeLayer::Position)
+                .ok_or_else(|| "No position reader available.")?,
+        );
+        let rgb_reader = BufReader::new(
+            layers
+                .remove(&NodeLayer::Color)
+                .ok_or_else(|| "No color reader available.")?,
+        );
+        let mut readers = vec![xyz_reader, rgb_reader];
+
+        if let Some(intensity_read) = layers.remove(&NodeLayer::Intensity) {
+            readers.push(BufReader::new(intensity_read));
+        };
+
+        Ok(Self {
+            readers,
+            position_encoding,
+            bounding_cube,
+        })
+    }
+}
+
+/// Streams points from our data provider representation.
+pub struct NodeIteratorWithData {
+    reader: Box<dyn NodeReader>,
```

Comment:
I'd make this a generic to avoid a dereference each time we read a point. Alternatively, we could pack this into the NodeReader trait (as two getter functions, which would hopefully get inlined), which would avoid an extra struct.

Reply:
With the latter option, we could also make the interface to NodeReader safer and have it return an Option<Point>, right?

Reply:
I like the suggestions, implemented all of them, PTAL!